prov/tcp/src/tcpx_cq.c

```diff
@@ -104,15 +104,29 @@ void tcpx_pe_entry_release(struct tcpx_pe_entry *pe_entry)
 	fastlock_release(&cq->cq_lock);
 }
 
-void tcpx_cq_report_completion(struct util_cq *cq,
-			       struct tcpx_pe_entry *pe_entry,
-			       int err)
+void tcpx_pe_entry_complete(struct tcpx_pe_entry *pe_entry,
+			    int err)
 {
+	struct util_cq *cq;
 	struct fi_cq_err_entry err_entry;
 
 	if (pe_entry->flags & TCPX_NO_COMPLETION)
 		return;
 
+	switch (pe_entry->msg_hdr.hdr.op_data) {
+	case TCPX_OP_MSG_SEND:
+	case TCPX_OP_WRITE:
+		cq = pe_entry->ep->util_ep.tx_cq;
+		break;
+	case TCPX_OP_MSG_RECV:
+	case TCPX_OP_REMOTE_WRITE:
+		cq = pe_entry->ep->util_ep.rx_cq;
+		break;
+	default:
+		FI_WARN(&tcpx_prov, FI_LOG_DOMAIN,"invalid pe_entry\n");
+		return;
+	}
```

Comment:
This introduces a switch statement, versus the previous code, where no branches were needed.  The older code should perform better.

Reply:
In case of  RMA Read, though process_rx_entry function receives the response data, the completion should be posted to tx_cq.  Similarly when FI_RMA_EVENT is enabled, RMA remote send response completion should be posted to rx_cq. Due to these cases, if statement would need to be added. 

One other way is to have the pointer to corresponding to cq in the entry itself, to avoid if or switch-case statements.

Reply:
Don't you know what the operation is at the time that the completion is being written?  That is, at least with all of the current code, we are working on a specific operation already, which makes this check unnecessary.  I wouldn't expect handling RMA to be any different -- we should know that we were handling an RMA read response, for example, prior to writing any completion.

Reply:
These commits are not structured in that way. RMA reads are trying to use the same progress rx entry. But it does make sense to break them up into separate functions.